Additional autograd unit tests for Python UDFs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29069 Fix distributed autograd initialization.
* **#29041 Additional autograd unit tests for Python UDFs.**

1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18273041](https://our.internmc.facebook.com/intern/diff/D18273041/)

Commit:
Additional autograd unit tests for Python UDFs.

1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)
ghstack-source-id: 92972494

Differential Revision: [D18273041](https://our.internmc.facebook.com/intern/diff/D18273041/)

[ghstack-poisoned]

pritamdamania87 commented:
@ezyang I'm not able to reproduce the doxygen issue locally. This is what I did and `check-doxygen.sh` completed successfully:

```
cd docs/
make html
cd cpp/source
./check-doxygen.sh
```

check-doxygen.sh output: 

```
++ dirname ./check-doxygen.sh
+ pushd ./../../..
~/local/pytorch ~/local/pytorch/docs/cpp/source
+ cp aten/src/ATen/common_with_cwrap.py tools/shared/cwrap_common.py
+ cp torch/_utils_internal.py tools/shared
+ python aten/src/ATen/gen.py -s aten/src/ATen -d build/aten/src/ATen aten/src/ATen/Declarations.cwrap aten/src/THNN/generic/THNN.h aten/src/THCUNN/generic/THCUNN.h aten/src/ATen/nn.yaml aten/src/ATen/native/native_functions.yaml
+ python tools/setup_helpers/generate_code.py --declarations-path build/aten/src/ATen/Declarations.yaml --nn-path aten/src
Skipped writing torch/csrc/autograd/generated/python_functions.h
Skipped writing torch/csrc/autograd/generated/python_functions.cpp
Skipped writing torch/csrc/autograd/generated/python_variable_methods.cpp
Skipped writing torch/csrc/autograd/generated/python_variable_methods_dispatch.h
Skipped writing torch/csrc/autograd/generated/python_torch_functions.cpp
Skipped writing torch/csrc/autograd/generated/python_torch_functions_dispatch.h
Skipped writing torch/csrc/autograd/generated/python_nn_functions.cpp
Skipped writing torch/csrc/autograd/generated/python_nn_functions.h
Skipped writing torch/csrc/autograd/generated/python_nn_functions_dispatch.h
Skipped writing torch/csrc/autograd/generated/VariableType.h
Skipped writing torch/csrc/autograd/generated/VariableType_0.cpp
Skipped writing torch/csrc/autograd/generated/VariableType_1.cpp
Skipped writing torch/csrc/autograd/generated/VariableType_2.cpp
Skipped writing torch/csrc/autograd/generated/VariableType_3.cpp
Skipped writing torch/csrc/autograd/generated/VariableType_4.cpp
Skipped writing torch/csrc/autograd/generated/VariableTypeEverything.cpp
Skipped writing torch/csrc/autograd/generated/Functions.h
Skipped writing torch/csrc/autograd/generated/Functions.cpp
Skipped writing torch/csrc/autograd/generated/variable_factories.h
Skipped writing torch/csrc/jit/generated/register_aten_ops_0.cpp
Skipped writing torch/csrc/jit/generated/register_aten_ops_1.cpp
Skipped writing torch/csrc/jit/generated/register_aten_ops_2.cpp
+ popd
~/local/pytorch/docs/cpp/source
+ doxygen
+ cp original-doxygen-log.txt doxygen-log.txt
+ ignore_warning 'warning: no uniquely matching class member found for'
+ set +e
+ grep -v 'warning: no uniquely matching class member found for' doxygen-log.txt
+ set -e
+ mv temp.txt doxygen-log.txt
+ ignore_warning 'warning: explicit link request to '\''Item'\'' could not be resolved'
+ set +e
+ grep -v 'warning: explicit link request to '\''Item'\'' could not be resolved' doxygen-log.txt
+ set -e
+ mv temp.txt doxygen-log.txt
+ ignore_warning 'warning: Included by graph for '\''types.h'\'' not generated, too many nodes'
+ set +e
+ grep -v 'warning: Included by graph for '\''types.h'\'' not generated, too many nodes' doxygen-log.txt
+ set -e
+ mv temp.txt doxygen-log.txt
++ grep warning: doxygen-log.txt
++ wc -l
+ warnings=0
+ echo 'Treating all remaining warnings as errors'
Treating all remaining warnings as errors
+ [[ 0 -ne 0 ]]
+ rm -f doxygen-log.txt original-doxygen-log.txt
```

ezyang commented:
This error may be sensitive to doxygen or dot version, and you may not have lined things up exactly the same way.

I was reading through the CI script and it looks like we filter out similar looking warnings. So it might be acceptable to just filter this one out too.

ezyang commented:
you could also do this https://github.com/pytorch/pytorch/pull/26613/files#diff-90c72f4e4e61cafd042a47b211ed4e05

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18273041](https://our.internmc.facebook.com/intern/diff/D18273041/)

[ghstack-poisoned]

pritamdamania87 commented:
> you could also do this https://github.com/pytorch/pytorch/pull/26613/files#diff-90c72f4e4e61cafd042a47b211ed4e05

Followed what Alban did there and quick_checks are now passing. Could you approve this PR? Thanks!

mrshenli reviewed (approved):
This is a retry of #28824, and the only difference is in `docs/cpp/source/Doxyfile`.

Lint failures on `python_engine.cpp` is irrelevant.
